baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/ThrowSpecificity.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.baseline.errorprone;
+
+import com.google.auto.service.AutoService;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterables;
+import com.google.errorprone.BugPattern;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.bugpatterns.BugChecker;
+import com.google.errorprone.fixes.SuggestedFix;
+import com.google.errorprone.fixes.SuggestedFixes;
+import com.google.errorprone.matchers.Description;
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.source.tree.ExpressionTree;
+import com.sun.source.tree.MethodTree;
+import com.sun.tools.javac.code.Symbol;
+import com.sun.tools.javac.code.Type;
+import com.sun.tools.javac.code.Types;
+import java.util.Collection;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import javax.lang.model.element.Modifier;
+
+@AutoService(BugChecker.class)
+@BugPattern(
+        name = "ThrowSpecificity",
+        link = "https://github.com/palantir/gradle-baseline#baseline-error-prone-checks",
+        linkType = BugPattern.LinkType.CUSTOM,
+        severity = BugPattern.SeverityLevel.SUGGESTION,
+        summary = "Prefer more specific error types than Exception and Throwable. When methods are updated to throw "
+                + "new checked exceptions they expect callers to handle failure types explicitly. Catching broad "
+                + "types defeats the type system. By catching the most specific types possible we leverage existing "
+                + "compiler functionality to detect unreachable code.")
+public final class ThrowSpecificity extends BugChecker implements BugChecker.MethodTreeMatcher {
```

Comment:
Also is this just a strictly better version of https://github.com/google/error-prone/blob/master/core/src/main/java/com/google/errorprone/bugpatterns/ThrowSpecificExceptions.java ?

Reply:
Oh wow is `state.errorProneOptions().isTestOnlyTarget()` new?

I think there's a lot of overlap, `ThrowSpecificExceptions.java` recommends changing `throw <throwable>` blocks to use more specific types, where this check produces no runtime changes, only provides stronger type safety by narrowing declared throws types on methods.